lib/openstack_query/runners/server_runner.py

```diff
@@ -37,89 +60,37 @@ def _run_query(
+        :param meta_params: a set of meta parameters that dictates how the query is run
 
         """
-        projects = self._get_projects(conn, from_projects)
-        logger.debug("running query on each project. Found %s projects", len(projects))
-        query_res = self._run_query_on_projects(conn, projects, filter_kwargs).values()
-        return [server for project_servers in query_res for server in project_servers]
+        if not filter_kwargs:
+            filter_kwargs = {}
 
-    def _get_projects(
-        self,
-        conn: OpenstackConnection,
-        projects: Optional[List[ProjectIdentifier]] = None,
-    ) -> List[Project]:
-        """
-        This method gets openstack projects from a list of project identifiers
-        :param conn: An OpenstackConnection object - used to connect to openstacksdk
-        :param projects: A list of project identifiers to get the associated openstack project object,
-        if None, gets all projects
-        """
-        if not projects:
-            logger.info("no projects given, query will run on all projects")
-            return list(conn.identity.projects())
-
-        all_projects = []
-        for project in projects:
-            if isinstance(project, Project):
-                all_projects.append(project)
-            else:
-                try:
-                    all_projects.append(
-                        conn.identity.find_project(project, ignore_missing=False)
-                    )
-                except ResourceNotFound as err:
-                    raise ParseQueryError(
-                        "Failed to execute query - could not find project to search in"
-                    ) from err
-        return all_projects
-
-    def _run_query_on_projects(
-        self,
-        conn: OpenstackConnection,
-        projects: List[Project],
-        filter_kwargs: Optional[Dict[str, str]] = None,
-    ) -> Dict[str, List[Server]]:
-        """
-        This method is a helper function that will run the query on a list of openstack projects given and return
-        a dictionary of servers that match the query grouped by project ids for which the servers belong to
-        :param conn: An OpenstackConnection object - used to connect to openstacksdk
-        :param projects: A list of openstacksdk projects to run query on
-        :param filter_kwargs: An Optional set of filter kwargs to pass to conn.compute.servers()
-        """
-        total = len(projects)
-        query_res = {}
-        for i, project in enumerate(projects):
-            logger.debug("running query on project %s / %s", i + 1, total)
-            query_res.update(
-                {
-                    project["id"]: self._run_query_on_project(
-                        conn, project, filter_kwargs
-                    )
-                }
+        # so we can search in all projects instead of default set in clouds.yaml
+        filter_kwargs.update({"all_tenants": True})
```

Comment:
This will only work with admin creds, can we only do it if we know we're in all projects mode effectively?

Reply:
pretty much every server-side filter is admin-only. Openstacksdk ignores any filter like "project_id" if you don't have admin rights - setting all_tenants here will actually produce an error rather than erroneous results

Reply:
plus `all_tenants` is required otherwise we only find servers for a single project that's mentioned in `clouds.yaml`